compiler/rustc_mir/src/transform/check_unsafety.rs

```diff
@@ -370,6 +388,56 @@ impl<'a, 'tcx> UnsafetyChecker<'a, 'tcx> {
             )
         }
     }
+
+    fn is_trait_upcasting_coercion_pointee(&self, source: Ty<'tcx>, target: Ty<'tcx>) -> bool {
+        match (source.kind(), target.kind()) {
+            (&ty::Dynamic(data_a, ..), &ty::Dynamic(data_b, ..)) => {
+                data_a.principal_def_id() != data_b.principal_def_id()
```

Comment:
Why is it okay to allow casts where `data_a.principal_def_id() == data_b.principal_def_id()`? Please add a comment.

Reply:
Ok!